src/creeps/Zombie.cpp

```diff
@@ -1,61 +1,313 @@
+
+    setCurDir(d);
+    if (state != ZOMBIE_MOVE) {
+        setState(ZOMBIE_MOVE);
+    }
 }
 
-void Zombie::generateRandomMove(){
-    double cosVal = cos(getAngle()*PI/180.0);
-    double sinVal = sin(getAngle()*PI/180.0);
+/**
+ * A* algo generates a string of direction digits.
+ * Fred Yang
+ * Feb 14
+ */
+string Zombie::generatePath(const float xStart, const float yStart,
+                            const float xDest, const float yDest) {
+    // priority queue index
+    int index = 0;
 
-    const int x = getVelocity();
-    const int y = getVelocity();
+    // temp index
+    int i, j;
 
-    const int tx = getX() + 20 * cosVal;
-    const int ty = getY() + 20 * sinVal;
+    // row / column index
+    int x, y, xdx, ydy;
 
-    movementHitBox->move(tx,ty);
-    projectileHitBox->move(tx,ty);
+    // temp char
+    char c;
 
-    CollisionHandler &ch = GameManager::instance()->getCollisionHandler();
-    if(ch.detectMovementCollision(this)){
-        setAngle(getRandomAngle());
-        cosVal = cos(getAngle() * PI / 180.0);
-        sinVal = sin(getAngle() * PI / 180.0);
+    // path to be generated
+    string path;
+
+    // current node & child node
+    static Node curNode;
+    static Node childNode;
+
+    // priority queue
+    static priority_queue<Node> pq[2];
+
+    // reset the node maps
+    for (i = 0; i < ROW; i++) {
+        for (j = 0; j < COL; j++) {
+            closedNodes[i][j] = 0;
+            openNodes[i][j] = 0;
+        }
+    }
+
+    int xNodeStart = (int) xStart / TILE_SIZE;
+    int yNodeStart = (int) yStart / TILE_SIZE;
+    int xNodeDest = (int) xDest / TILE_SIZE;
+    int yNodeDest = (int) yDest / TILE_SIZE;
+
+    // create the start node and push into open list
+    curNode = Node(xNodeStart, yNodeStart, 0, 0);
+    curNode.updatePriority(xNodeDest, yNodeDest);
+    pq[index].push(curNode);
+
+    // A* path finding
+    while (!pq[index].empty()) {
+        // get the current node with the highest priority from open list
+        curNode = Node(pq[index].top().getXPos(), pq[index].top().getYPos(),
+                           pq[index].top().getLevel(), pq[index].top().getPriority());
+
+        x = curNode.getXPos();
+        y = curNode.getYPos();
+
+        // remove the node from the open list
+        pq[index].pop();
+        openNodes[x][y] = 0;
+
+        // mark it on the closed nodes map
+        closedNodes[x][y] = 1;
+
+        // quit searching when the destination is reached
+        if (x == xNodeDest && y == yNodeDest) {
+            // generate the path from destination to start
+            // by following the directions
+            path = "";
+            while (!(x == xNodeStart && y == yNodeStart)) {
+                j = dirMap[x][y];
+                c = '0' + (j + DIR_CAP/2)%DIR_CAP;
+                path = c + path;
+                x += MX[j];
+                y += MY[j];
+            }
+
+            // empty the leftover nodes
+            while (!pq[index].empty()) {
```

Comment:
This can be changed into something cleaner
Use `pq[index] = std::priorityQueue<Node()`